crates/s3_vectors/src/table.rs

```diff
@@ -0,0 +1,268 @@
+use std::{any::Any, fmt, sync::Arc};
+
+use async_trait::async_trait;
+use aws_sdk_s3vectors::types::{Index, VectorData};
+use datafusion::{
+    arrow::{
+        array::{Array, Float32Array, ListArray},
+        datatypes::{DataType, Field, Schema, SchemaRef},
+    },
+    catalog::{Session, TableProvider, memory::DataSourceExec},
+    common::{
+        Result,
+        tree_node::{TreeNode, TreeNodeRecursion},
+    },
+    datasource::{TableType, sink::DataSinkExec, source::DataSource},
+    logical_expr::{Expr, TableProviderFilterPushDown, dml::InsertOp, expr::ScalarFunction},
+    physical_plan::ExecutionPlan,
+    scalar::ScalarValue,
+};
+
+use crate::{
+    S3Vectors, list_vectors::ListVectorsSource, put_vectors::PutVectorsSink,
+    query_vectors::QueryVectorsSource,
+};
+
+static NAME: &str = "S3VectorsTable";
+
+const DEFAULT_TOP_K: usize = 10;
+
+pub struct S3VectorsTable {
+    client: Arc<dyn S3Vectors + Send + Sync>,
+    index: Index,
+    schema: SchemaRef,
+}
+
+impl S3VectorsTable {
+    pub fn new(client: Arc<dyn S3Vectors + Send + Sync>, index: Index) -> Self {
+        let schema = Arc::new(Schema::new(vec![
+            Field::new("key", DataType::Utf8, false),
+            Field::new(
+                "data",
+                DataType::List(Arc::new(Field::new("item", DataType::Float32, true))),
+                true,
+            ),
+            Field::new("distance", DataType::Float32, true),
+        ]));
+        Self {
+            client,
+            index,
+            schema,
+        }
+    }
+}
+
+impl fmt::Debug for S3VectorsTable {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{NAME} bucket={} index={}",
+            self.index.vector_bucket_name(),
+            self.index.index_name
+        )
+    }
+}
+
+#[async_trait]
+impl TableProvider for S3VectorsTable {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        _state: &dyn Session,
+        _projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        let mut source = None;
+        for filter in filters {
```

Comment:
Do we support multiple predicates? Either way is fine: clear error if we don't, or union if we do?